jolie/src/main/java/jolie/runtime/typing/BasicType.java

```diff
@@ -81,6 +81,15 @@ public NativeType nativeType() {
 		return nativeType;
 	}
 
+	@SuppressWarnings( "unchecked" )
+	private static < T > void checkRefinement( Object nativeValue, BasicTypeRefinement< T > refinement,
```

Comment:
Why not make the type of `nativeValue` a `T`?

Reply:
This did not work by means the `for` loop:
```
			for( BasicTypeRefinement< ? > refinement : refinements )
				checkRefinement( nativeValue, refinement, pathSupplier );
```

On one place you need to weaken the generics, otherwise the underlying issue may not be solved (to have this combined data-type with two kinds of generics: the basic type with the refinements of some other type).

Reply:
E.g. you could do this, but it is not much better:
```
	@SuppressWarnings( "unchecked" )
	private static < T,U > void checkRefinement( U nativeValue, BasicTypeRefinement< T > refinement,
		Supplier< String > pathSupplier ) throws TypeCheckingException {
		if( !refinement.checkValue( ( T ) nativeValue ) ) {
...
```

Reply:
Any thoughts?

Reply:
What if you add the constraints 'U extends T' in the generic parameter? Does that help a bit maybe? (This might require updating the signature of the refinements as well)

Reply:
The problem here is that you are not able to solve it completely by not making also `BasicTypeDefinition` generic for the refinements which I wanted to avoid. And also then I guess that the discrepancy remains:
```
public static BasicType< ? > fromBasicTypeDefinition( BasicTypeDefinition basicTypeDefinition ) {
	return fromBasicTypeDefinitionInternal( PURE_BASIC_TYPES.get( basicTypeDefinition.nativeType() ),
			basicTypeDefinition.refinements() );
}
```
We are not able to concile the `List< BasicTypeRefinement< ? super T > >` for a certain basic type which we grasp from the map `PURE_BASIC_TYPES` with some additional `List< BasicTypeRefinement< ? > >` from the `refinements()` method.

So I have opted for creating this `fromBasicTypeDefinitionRefinements()` method which allocates a new list and passes the entries one by one. I could have made it also with a whole-list cast `(List<BasicTypeRefinement< ? super T >>)((List) ...)` to avoid the additional allocation but it didn't look like a very clean approach.